### Aula 1 - Importar os dados

Chegou a hora, Pietro! Você vai começar a sua jornada no #7DaysOfCode.

Estou muito animado para começar esse desafio! Iremos explorar os dados de empréstimos dos acervos do sistema de bibliotecas da UFRN.

Um dos objetivos de uma biblioteca é garantir que os materiais informacionais estejam sendo utilizados. Os empréstimos realizados podem ser um indicador, mesmo que de forma básica (pois você não consegue garantir que haja uma leitura ou utilização real).

Por este motivo, entender a quantidade de empréstimos se torna importante.

Questões de diferentes perspectivas podem surgir como:


A quantidade de empréstimos está aumentando ou diminuindo ao decorrer dos últimos anos?
Em quais bibliotecas do sistema estão a maior quantidade de empréstimos?
Quais são os temas mais emprestados? E os menos?

Com estas e outras informações será possível entender o cenário e apresentá-lo à diretoria das bibliotecas, para que possam tomar melhores decisões na melhoria da infraestrutura, dos recursos e processos da unidade de informação.

Mas para que tudo isso seja realizado, você precisará começar com a coleta e organização dos dados para que possa trabalhar com eles nas próximas análises.

Borá lá?!

Você trabalhará com dados apenas dos últimos 10 anos disponíveis. Por isso, importe para seu Jupyter Notebook os dados de:


Empréstimos dos acervos das bibliotecas de UFRN
Exemplares do acervo

Dados baixados? Ok, mas são diversas tabelas diferentes e isso dificulta o trabalho. Portanto, o seu primeiro passo é unificar em um único Dataframe todos os dados pertinentes para a análise.

Comece pelos empréstimos e você terá os dados das transações. Depois, mescle com os dados do acervo, para que você possa entender, por exemplo, de qual biblioteca era o material emprestado ou a qual tema ele se referia. Elas se relacionam pela coluna de código de barras de cada material.

Lembre-se que é muito comum receber dados nulos ou duplicados, por isso não deixe de fazer a limpeza.

In [1]:
import pandas as pd

## Dados Empréstimos

In [3]:
url = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20201.csv'

emprestimos = pd.read_csv(url)

print(emprestimos.head())

   id_emprestimo codigo_barras                 data_renovacao  \
0        2962509    2007005628                            NaN   
1        2962510       L202802                            NaN   
2        2962511    2012008328                            NaN   
3        2962512    2012055016  2020/01/17 09:56:18.380000000   
4        2962513       L197391  2020/01/17 23:50:18.862000000   

                 data_emprestimo                 data_devolucao  \
0  2020/01/02 08:17:30.290000000  2020/01/07 11:14:07.000000000   
1  2020/01/02 08:17:30.391000000  2020/01/07 11:14:17.000000000   
2  2020/01/02 08:17:30.436000000  2020/01/07 11:24:46.000000000   
3  2020/01/02 08:36:40.163000000  2020/01/31 14:53:52.000000000   
4  2020/01/02 09:36:03.877000000  2020/01/24 08:18:19.000000000   

   matricula_ou_siape    tipo_vinculo_usuario  
0        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
1        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
2        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
3        2.019

In [2]:
url_20201 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20201.csv'
url_20192 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20192.csv'
url_20191 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20191.csv'
url_20182 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20182.csv'
url_20181 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20181.csv'
url_20172 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20172.csv'
url_20171 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20171.csv'
url_20162 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20162.csv'
url_20161 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20161.csv'
url_20152 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20152.csv'
url_20151 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20151.csv'
url_20142 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20142.csv'
url_20141 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20141.csv'
url_20132 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20132.csv'
url_20131 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20131.csv'
url_20122 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20122.csv'
url_20121 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20121.csv'
url_20112 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20112.csv'
url_20111 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20111.csv'
url_20102 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20102.csv'
url_20101 = 'https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_1-Importando_dados/Datasets/dados_emprestimos/emprestimos-20101.csv'


urls = [
    url_20201, url_20192, url_20191, url_20182, url_20181,
    url_20172, url_20171, url_20162, url_20161, url_20152,
    url_20151, url_20142, url_20141, url_20132, url_20131,
    url_20122, url_20121, url_20112, url_20111, url_20102,
    url_20101
]

# Carregar todos os DataFrames em uma lista
emprestimos = [pd.read_csv(url) for url in urls]

# Concatenar todos os DataFrames em um único DataFrame
todos_emprestimos = pd.concat(emprestimos, ignore_index=True)

print(todos_emprestimos.head())


   id_emprestimo codigo_barras                 data_renovacao  \
0        2962509    2007005628                            NaN   
1        2962510       L202802                            NaN   
2        2962511    2012008328                            NaN   
3        2962512    2012055016  2020/01/17 09:56:18.380000000   
4        2962513       L197391  2020/01/17 23:50:18.862000000   

                 data_emprestimo                 data_devolucao  \
0  2020/01/02 08:17:30.290000000  2020/01/07 11:14:07.000000000   
1  2020/01/02 08:17:30.391000000  2020/01/07 11:14:17.000000000   
2  2020/01/02 08:17:30.436000000  2020/01/07 11:24:46.000000000   
3  2020/01/02 08:36:40.163000000  2020/01/31 14:53:52.000000000   
4  2020/01/02 09:36:03.877000000  2020/01/24 08:18:19.000000000   

   matricula_ou_siape    tipo_vinculo_usuario  
0        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
1        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
2        2.018103e+10  ALUNO DE PÓS-GRADUAÇÃO  
3        2.019

In [3]:
emprestimos_totais = pd.DataFrame(todos_emprestimos)

In [4]:
print(emprestimos_totais.dtypes)

id_emprestimo             int64
codigo_barras            object
data_renovacao           object
data_emprestimo          object
data_devolucao           object
matricula_ou_siape      float64
tipo_vinculo_usuario     object
dtype: object


In [5]:
emprestimos_totais.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2258018 entries, 0 to 2258017
Data columns (total 7 columns):
 #   Column                Dtype  
---  ------                -----  
 0   id_emprestimo         int64  
 1   codigo_barras         object 
 2   data_renovacao        object 
 3   data_emprestimo       object 
 4   data_devolucao        object 
 5   matricula_ou_siape    float64
 6   tipo_vinculo_usuario  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 120.6+ MB


In [6]:
emprestimos_totais['data_renovacao'] = pd.to_datetime(emprestimos_totais['data_renovacao'], errors='coerce')
emprestimos_totais['data_emprestimo'] = pd.to_datetime(emprestimos_totais['data_emprestimo'], errors='coerce')
emprestimos_totais['data_devolucao'] = pd.to_datetime(emprestimos_totais['data_devolucao'], errors='coerce')

In [7]:
emprestimos_totais.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2258018 entries, 0 to 2258017
Data columns (total 7 columns):
 #   Column                Dtype         
---  ------                -----         
 0   id_emprestimo         int64         
 1   codigo_barras         object        
 2   data_renovacao        datetime64[ns]
 3   data_emprestimo       datetime64[ns]
 4   data_devolucao        datetime64[ns]
 5   matricula_ou_siape    float64       
 6   tipo_vinculo_usuario  object        
dtypes: datetime64[ns](3), float64(1), int64(1), object(2)
memory usage: 120.6+ MB


In [8]:
emprestimos_totais.head()

id_emprestimo codigo_barras          data_renovacao  \
0        2962509    2007005628                     NaT   
1        2962510       L202802                     NaT   
2        2962511    2012008328                     NaT   
3        2962512    2012055016 2020-01-17 09:56:18.380   
4        2962513       L197391 2020-01-17 23:50:18.862   

          data_emprestimo      data_devolucao  matricula_ou_siape  \
0 2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1 2020-01-02 08:17:30.391 2020-01-07 11:14:17        2.018103e+10   
2 2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
3 2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
4 2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   

     tipo_vinculo_usuario  
0  ALUNO DE PÓS-GRADUAÇÃO  
1  ALUNO DE PÓS-GRADUAÇÃO  
2  ALUNO DE PÓS-GRADUAÇÃO  
3      ALUNO DE GRADUAÇÃO  
4      ALUNO DE GRADUAÇÃO

In [9]:
nulos_por_coluna = emprestimos_totais.isnull().sum()
print(nulos_por_coluna)

id_emprestimo                 0
codigo_barras                 0
data_renovacao          1285720
data_emprestimo               0
data_devolucao             6471
matricula_ou_siape         3170
tipo_vinculo_usuario          0
dtype: int64


In [10]:
emprestimos_totais.value_counts()

id_emprestimo  codigo_barras  data_renovacao           data_emprestimo          data_devolucao           matricula_ou_siape  tipo_vinculo_usuario  
709705         2009032168     2010-01-17 18:56:43.909  2010-01-04 09:58:46.549  2010-01-20 13:14:34.464  2.008010e+09        ALUNO DE GRADUAÇÃO        7
709704         2006021442     2010-01-17 18:56:43.923  2010-01-04 09:58:46.522  2010-02-01 13:59:02.528  2.008010e+09        ALUNO DE GRADUAÇÃO        7
709698         2009047725     2010-01-18 14:44:41.163  2010-01-04 09:21:19.099  2010-02-02 12:02:38.444  2.009047e+09        ALUNO DE GRADUAÇÃO        3
709686         2006017618     2010-01-26 08:07:01.738  2010-01-04 08:08:44.081  2010-02-25 07:36:25.800  2.008112e+09        ALUNO DE PÓS-GRADUAÇÃO    2
709687         L184117        2010-01-18 11:07:46.470  2010-01-04 08:24:21.284  2010-02-03 08:58:45.692  2.007211e+08        ALUNO DE GRADUAÇÃO        2
                                                                                       

In [11]:
emprestimos_totais.drop_duplicates()

id_emprestimo codigo_barras          data_renovacao  \
0              2962509    2007005628                     NaT   
1              2962510       L202802                     NaT   
2              2962511    2012008328                     NaT   
3              2962512    2012055016 2020-01-17 09:56:18.380   
4              2962513       L197391 2020-01-17 23:50:18.862   
...                ...           ...                     ...   
2258013         919377       C019571                     NaT   
2258014         919379       C019573                     NaT   
2258015         919381       C020714                     NaT   
2258016         919382       C020845                     NaT   
2258017         919431       Y011015                     NaT   

                data_emprestimo      data_devolucao  matricula_ou_siape  \
0       2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1       2020-01-02 08:17:30.391 2020-01-07 11:14:17        2.018103e+10   
2       2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
3       2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
4       2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   
...                         ...                 ...                 ...   
2258013 2010-05-28 13:12:45.000 2010-05-28 17:42:02        1.687161e+06   
2258014 2010-03-05 13:26:50.000 2010-04-08 18:52:47        1.687161e+06   
2258015 2010-06-17 20:30:58.000 2010-06-24 10:51:23        1.687161e+06   
2258016 2010-03-11 15:08:05.000 2010-03-11 17:24:24        1.687161e+06   
2258017 2010-06-17 08:12:48.000 2010-06-30 07:34:08        2.010204e+09   

           tipo_vinculo_usuario  
0        ALUNO DE PÓS-GRADUAÇÃO  
1        ALUNO DE PÓS-GRADUAÇÃO  
2        ALUNO DE PÓS-GRADUAÇÃO  
3            ALUNO DE GRADUAÇÃO  
4            ALUNO DE GRADUAÇÃO  
...                         ...  
2258013                 DOCENTE  
2258014                 DOCENTE  
2258015                 DOCENTE  
2258016                 DOCENTE  
2258017  ALUNO DE PÓS-GRADUAÇÃO  

[2257981 rows x 7 columns]

## Dados Exemplares

In [12]:
dados_exemplares = pd.read_parquet('data/dados_exemplares.parquet')
dados_exemplares

id_exemplar codigo_barras            colecao  \
index                                                  
0                 5       L000003  Acervo Circulante   
1                 4       L000002  Acervo Circulante   
2                 3       L000001  Acervo Circulante   
3                 7       L000114  Acervo Circulante   
5                10       L000041  Acervo Circulante   
...             ...           ...                ...   
568580      1353472    2021011150  Acervo Circulante   
568581      1353473    2019013454  Acervo Circulante   
568582      1353474    2019012811  Acervo Circulante   
568583      1353475    2019013451  Acervo Circulante   
568584      1353476    2019012276  Acervo Circulante   

                                               biblioteca status_material  \
index                                                                       
0                          Biblioteca Central Zila Mamede         REGULAR   
1                          Biblioteca Central Zila Mamede         REGULAR   
2                          Biblioteca Central Zila Mamede        ESPECIAL   
3                          Biblioteca Central Zila Mamede         REGULAR   
5                          Biblioteca Central Zila Mamede        ESPECIAL   
...                                                   ...             ...   
568580  Biblioteca Setorial do Núcleo de Educação da I...         REGULAR   
568581                     Biblioteca Central Zila Mamede        ESPECIAL   
568582                     Biblioteca Central Zila Mamede        ESPECIAL   
568583                     Biblioteca Central Zila Mamede        ESPECIAL   
568584                     Biblioteca Central Zila Mamede        ESPECIAL   

        localizacao  registro_sistema  
index                                  
0               694                 1  
1               688                 1  
2               638                 1  
3               616                 5  
5               657                15  
...             ...               ...  
568580          951            268231  
568581          997            268233  
568582          987            268234  
568583          764            268235  
568584          962            268236  

[546237 rows x 7 columns]

In [13]:
dados_exemplares.value_counts()

id_exemplar  codigo_barras                colecao            biblioteca                      status_material  localizacao  registro_sistema
1353476      2019012276                   Acervo Circulante  Biblioteca Central Zila Mamede  ESPECIAL         962          268236              1
3            L000001                      Acervo Circulante  Biblioteca Central Zila Mamede  ESPECIAL         638          1                   1
4            L000002                      Acervo Circulante  Biblioteca Central Zila Mamede  REGULAR          688          1                   1
5            L000003                      Acervo Circulante  Biblioteca Central Zila Mamede  REGULAR          694          1                   1
7            L000114                      Acervo Circulante  Biblioteca Central Zila Mamede  REGULAR          616          5                   1
                                                                                                                                       

In [14]:
nulos_por_coluna = dados_exemplares.isnull().sum()
print(nulos_por_coluna)

id_exemplar         0
codigo_barras       0
colecao             0
biblioteca          0
status_material     0
localizacao         0
registro_sistema    0
dtype: int64


## Unindo os DataFrames

In [15]:
emprestimos_completos = emprestimos_totais.merge(dados_exemplares)
emprestimos_completos

id_emprestimo codigo_barras          data_renovacao  \
0              2962509    2007005628                     NaT   
1              2962511    2012008328                     NaT   
2              2962512    2012055016 2020-01-17 09:56:18.380   
3              2962513       L197391 2020-01-17 23:50:18.862   
4              2962514    2016030245 2020-01-17 23:50:18.835   
...                ...           ...                     ...   
2072553         799095    2007024245                     NaT   
2072554         799096       L197893                     NaT   
2072555         799097       L194357                     NaT   
2072556         799098    2010002179 2010-07-15 08:20:13.242   
2072557         799099    2009059990 2010-07-15 08:20:13.222   

                data_emprestimo      data_devolucao  matricula_ou_siape  \
0       2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1       2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
2       2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
3       2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   
4       2020-01-02 09:36:03.938 2020-01-24 08:18:13        2.018005e+10   
...                         ...                 ...                 ...   
2072553 2010-06-30 21:32:14.476 2010-07-02 21:16:43        2.007117e+08   
2072554 2010-06-30 21:32:14.618 2010-07-07 19:25:13        2.007117e+08   
2072555 2010-06-30 21:32:40.359 2010-07-01 16:51:54        2.009009e+09   
2072556 2010-06-30 21:40:41.356 2010-07-30 11:10:38        2.004186e+08   
2072557 2010-06-30 21:40:41.404 2010-08-19 14:34:15        2.004186e+08   

           tipo_vinculo_usuario  id_exemplar            colecao  \
0        ALUNO DE PÓS-GRADUAÇÃO       216276  Acervo Circulante   
1        ALUNO DE PÓS-GRADUAÇÃO       971301  Acervo Circulante   
2            ALUNO DE GRADUAÇÃO      1019380  Acervo Circulante   
3            ALUNO DE GRADUAÇÃO       177552  Acervo Circulante   
4            ALUNO DE GRADUAÇÃO      1243242  Acervo Circulante   
...                         ...          ...                ...   
2072553      ALUNO DE GRADUAÇÃO       223127  Acervo Circulante   
2072554      ALUNO DE GRADUAÇÃO       178672  Acervo Circulante   
2072555      ALUNO DE GRADUAÇÃO       164543  Acervo Circulante   
2072556      ALUNO DE GRADUAÇÃO       767241  Acervo Circulante   
2072557      ALUNO DE GRADUAÇÃO       286952  Acervo Circulante   

                             biblioteca status_material  localizacao  \
0        Biblioteca Central Zila Mamede         REGULAR          640   
1        Biblioteca Central Zila Mamede         REGULAR          350   
2        Biblioteca Central Zila Mamede         REGULAR          696   
3        Biblioteca Central Zila Mamede         REGULAR          699   
4        Biblioteca Central Zila Mamede         REGULAR          508   
...                                 ...             ...          ...   
2072553  Biblioteca Central Zila Mamede         REGULAR          655   
2072554  Biblioteca Central Zila Mamede         REGULAR          639   
2072555  Biblioteca Central Zila Mamede         REGULAR          657   
2072556  Biblioteca Central Zila Mamede         REGULAR          649   
2072557  Biblioteca Central Zila Mamede         REGULAR          655   

         registro_sistema  
0                   87626  
1                  176661  
2                  168729  
3                   67693  
4                  235780  
...                   ...  
2072553             91066  
2072554             68239  
2072555             61317  
2072556             89936  
2072557             68021  

[2072558 rows x 13 columns]

### Dia 2 - Limpeza de Dados

Hoje é o segundo dia do seu desafio #7DaysOfCode com Python Pandas, bora praticar?

Ontem você importou e organizou os dados. Hoje, você irá começar a manipulá-los, ou seja, tirar o que não for necessário, agrupar dados, atribuir novas informações, etc.

Claro, os dados precisam fazer sentido e estar preparados para o contexto da análise a ser realizada. Por isso, a “limpeza” dos dados é uma parte essencial em um projeto de ciência de dados.

Você irá iniciar a limpeza e atribuir mais contexto aos seus dados para depois aprofundar-se nas análises.

A Ciência de Dados é uma área interdisciplinar que abrange programação, matemática/estatística e conhecimento do negócio. Seu objetivo é extrair de dados, informações úteis que agregam valor e resolvem problemas. Entender o negócio é fundamental para que não se tenha análises sem explicações ou mesmo sem o foco necessário para a resolução dos problemas.

Você deve ter visto que tem uma coluna identificada como “localização” e diversos números nela, mas você sabe o que significam estes números?


"Os itens do acervo em uma biblioteca são organizados por um sistema de classificação de acordo com o respectivo tema. Existem diversos sistemas, mas este conjunto está de acordo com a CDU - Classificação Decimal Universal. Esta classificação é decimal, pois varia de acordo com a classe de cada assunto:
 

000 a 099: Generalidades. Ciência e conhecimento.
100 a 199: Filosofia e psicologia.
200 a 299: Religião.
300 a 399: Ciências sociais.
400 a 499: Classe vaga. Provisoriamente não ocupada.
500 a 599: Matemática e ciências naturais.
600 a 699: Ciências aplicadas.
700 a 799: Belas artes.
800 a 899: Linguagem. Língua. Linguística.
900 a 999: Geografia. Biografia. História."

Portanto, se um material tiver um código de localização 720, ele está dentro da classe geral de “Belas Artes”; ou se tiver um código 028, estará dentro da classe geral de “Generalidades. Ciência e conhecimento”.

Para isso, crie uma nova coluna com os valores da localização, para refletir a respectiva classe geral na CDU.

Você precisará ainda excluir alguns dados e modificar outros.  

A coluna "registro_sistema", por exemplo, não está fazendo sentido para essa análise, por isso você pode exclui-la.
Já a coluna da matricula (“matricula_ou_siape”) não está com um formato muito legível. Transforme-a em formato String.

In [16]:
emprestimos_completos.head()

id_emprestimo codigo_barras          data_renovacao  \
0        2962509    2007005628                     NaT   
1        2962511    2012008328                     NaT   
2        2962512    2012055016 2020-01-17 09:56:18.380   
3        2962513       L197391 2020-01-17 23:50:18.862   
4        2962514    2016030245 2020-01-17 23:50:18.835   

          data_emprestimo      data_devolucao  matricula_ou_siape  \
0 2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1 2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
2 2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
3 2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   
4 2020-01-02 09:36:03.938 2020-01-24 08:18:13        2.018005e+10   

     tipo_vinculo_usuario  id_exemplar            colecao  \
0  ALUNO DE PÓS-GRADUAÇÃO       216276  Acervo Circulante   
1  ALUNO DE PÓS-GRADUAÇÃO       971301  Acervo Circulante   
2      ALUNO DE GRADUAÇÃO      1019380  Acervo Circulante   
3      ALUNO DE GRADUAÇÃO       177552  Acervo Circulante   
4      ALUNO DE GRADUAÇÃO      1243242  Acervo Circulante   

                       biblioteca status_material  localizacao  \
0  Biblioteca Central Zila Mamede         REGULAR          640   
1  Biblioteca Central Zila Mamede         REGULAR          350   
2  Biblioteca Central Zila Mamede         REGULAR          696   
3  Biblioteca Central Zila Mamede         REGULAR          699   
4  Biblioteca Central Zila Mamede         REGULAR          508   

   registro_sistema  
0             87626  
1            176661  
2            168729  
3             67693  
4            235780

In [17]:
CDU_lista = []
for CDU in emprestimos_completos['localizacao']:
    if CDU < 100:
        CDU_lista.append("Generalidades. Ciência e conhecimento.")
    elif CDU < 200:
        CDU_lista.append("Filosofia e psicologia.")
    elif CDU < 300:
        CDU_lista.append("Religião.")
    elif CDU < 400:
        CDU_lista.append("Ciências sociais.")
    elif CDU < 500:
        CDU_lista.append("Classe vaga. Provisoriamente não ocupada.")
    elif CDU < 600:
        CDU_lista.append("Matemática e ciências naturais.")
    elif CDU < 700:
        CDU_lista.append("Ciências aplicadas.")
    elif CDU < 800:
        CDU_lista.append("Belas artes.")
    elif CDU < 900:
        CDU_lista.append("Linguagem. Língua. Linguística.")
    elif CDU < 1000:
        CDU_lista.append("Geografia. Biografia. História.")

emprestimos_completos['CDU_geral'] = CDU_lista


In [18]:
emprestimos_completos.head()

id_emprestimo codigo_barras          data_renovacao  \
0        2962509    2007005628                     NaT   
1        2962511    2012008328                     NaT   
2        2962512    2012055016 2020-01-17 09:56:18.380   
3        2962513       L197391 2020-01-17 23:50:18.862   
4        2962514    2016030245 2020-01-17 23:50:18.835   

          data_emprestimo      data_devolucao  matricula_ou_siape  \
0 2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1 2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
2 2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
3 2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   
4 2020-01-02 09:36:03.938 2020-01-24 08:18:13        2.018005e+10   

     tipo_vinculo_usuario  id_exemplar            colecao  \
0  ALUNO DE PÓS-GRADUAÇÃO       216276  Acervo Circulante   
1  ALUNO DE PÓS-GRADUAÇÃO       971301  Acervo Circulante   
2      ALUNO DE GRADUAÇÃO      1019380  Acervo Circulante   
3      ALUNO DE GRADUAÇÃO       177552  Acervo Circulante   
4      ALUNO DE GRADUAÇÃO      1243242  Acervo Circulante   

                       biblioteca status_material  localizacao  \
0  Biblioteca Central Zila Mamede         REGULAR          640   
1  Biblioteca Central Zila Mamede         REGULAR          350   
2  Biblioteca Central Zila Mamede         REGULAR          696   
3  Biblioteca Central Zila Mamede         REGULAR          699   
4  Biblioteca Central Zila Mamede         REGULAR          508   

   registro_sistema                        CDU_geral  
0             87626              Ciências aplicadas.  
1            176661                Ciências sociais.  
2            168729              Ciências aplicadas.  
3             67693              Ciências aplicadas.  
4            235780  Matemática e ciências naturais.

In [19]:
emprestimos_completos.drop(columns=['registro_sistema'],inplace=True)

In [20]:
emprestimos_totais['matricula_ou_siape'] = emprestimos_totais['matricula_ou_siape'].astype('string')

In [21]:
emprestimos_completos.head()

id_emprestimo codigo_barras          data_renovacao  \
0        2962509    2007005628                     NaT   
1        2962511    2012008328                     NaT   
2        2962512    2012055016 2020-01-17 09:56:18.380   
3        2962513       L197391 2020-01-17 23:50:18.862   
4        2962514    2016030245 2020-01-17 23:50:18.835   

          data_emprestimo      data_devolucao  matricula_ou_siape  \
0 2020-01-02 08:17:30.290 2020-01-07 11:14:07        2.018103e+10   
1 2020-01-02 08:17:30.436 2020-01-07 11:24:46        2.018103e+10   
2 2020-01-02 08:36:40.163 2020-01-31 14:53:52        2.019015e+10   
3 2020-01-02 09:36:03.877 2020-01-24 08:18:19        2.018005e+10   
4 2020-01-02 09:36:03.938 2020-01-24 08:18:13        2.018005e+10   

     tipo_vinculo_usuario  id_exemplar            colecao  \
0  ALUNO DE PÓS-GRADUAÇÃO       216276  Acervo Circulante   
1  ALUNO DE PÓS-GRADUAÇÃO       971301  Acervo Circulante   
2      ALUNO DE GRADUAÇÃO      1019380  Acervo Circulante   
3      ALUNO DE GRADUAÇÃO       177552  Acervo Circulante   
4      ALUNO DE GRADUAÇÃO      1243242  Acervo Circulante   

                       biblioteca status_material  localizacao  \
0  Biblioteca Central Zila Mamede         REGULAR          640   
1  Biblioteca Central Zila Mamede         REGULAR          350   
2  Biblioteca Central Zila Mamede         REGULAR          696   
3  Biblioteca Central Zila Mamede         REGULAR          699   
4  Biblioteca Central Zila Mamede         REGULAR          508   

                         CDU_geral  
0              Ciências aplicadas.  
1                Ciências sociais.  
2              Ciências aplicadas.  
3              Ciências aplicadas.  
4  Matemática e ciências naturais.